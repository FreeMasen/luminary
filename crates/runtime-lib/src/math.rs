#[cfg(feature = "runtime")]
use crate::{tags, TValue, TValueInner};
#[cfg(feature = "runtime")]
use core::ops::{Add, Mul, Sub};

#[cfg(feature = "runtime")]
macro_rules! bin_math_prefix {
    ($l:ident, $r:ident, $o:ident) => {
        (
            $crate::get_or_return!($l, $o),
            $crate::get_or_return!($r, $o),
            $crate::get_mut_or_return!($o),
        )
    };
}
#[cfg(feature = "runtime")]
macro_rules! bin_math_op {
    ($l:ident, $r:ident, $o:ident, $iop:path, $fop:path) => {{
        let (l, r, o) = bin_math_prefix!($l, $r, $o);

        *o = match (l.tag, r.tag) {
            (tags::INTEGER, tags::INTEGER) => TValue::new_int($iop(l.value.i, r.value.i)),
            (tags::FLOAT, tags::FLOAT) => {
                TValue::new_float($fop(l.value.f as f64, r.value.f as f64))
            }
            (tags::INTEGER, tags::FLOAT) => TValue::new_float($fop(l.value.i as f64, r.value.f)),
            (tags::FLOAT, tags::INTEGER) => TValue::new_float($fop(l.value.f, r.value.i as f64)),
            _ => TValue::new_bool(false),
        };
    }};
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn add(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    bin_math_op!(lhs, rhs, out, i64::wrapping_add, f64::add)
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn sub(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    bin_math_op!(lhs, rhs, out, i64::wrapping_sub, f64::sub)
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn mul(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    bin_math_op!(lhs, rhs, out, i64::wrapping_mul, f64::mul)
}

#[cfg(feature = "runtime")]
fn int_rem(lhs: i64, rhs: i64) -> TValue {
    if rhs == 0 {
        return crate::TValue::new_bool(false);
    }
    let Some(mut ret_val) = lhs.checked_rem_euclid(rhs) else {
        return float_rem(lhs as _, rhs as _);
    };
    // see lua5.3 lvm.c#568
    if ret_val != 0 && ret_val ^ rhs < 0 {
        ret_val += rhs;
    }
    crate::TValue::new_int(ret_val)
}

#[cfg(feature = "runtime")]
fn float_rem(lhs: f64, rhs: f64) -> TValue {
    let mut ret_val = lhs % rhs;
    // see lua5.3 llimits.h:280
    if ret_val * rhs < 0.0 {
        ret_val += rhs;
    }
    crate::TValue::new_float(ret_val)
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn rem(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    let (l, r, o) = bin_math_prefix!(lhs, rhs, out);
    let (l, r) = unsafe {
        match (l.tag, r.tag) {
            (tags::INTEGER, tags::INTEGER) => {
                *o = int_rem(l.value.i, r.value.i);
                return;
            }
            (tags::INTEGER, tags::FLOAT) => (l.value.i as f64, r.value.f),
            (tags::FLOAT, tags::INTEGER) => (l.value.f, r.value.i as f64),
            (tags::FLOAT, tags::FLOAT) => (l.value.f, r.value.f),
            _ => {
                *o = TValue::new_bool(false);
                return;
            }
        }
    };
    *o = float_rem(l, r)
}

#[cfg(feature = "runtime")]
fn tvalue_div_(l: &TValue, r: &TValue, o: &mut TValue) {
    let (l, r) = unsafe {
        match (l.tag, r.tag) {
            (tags::INTEGER, tags::INTEGER) => (l.value.i as f64, r.value.i as f64),
            (tags::INTEGER, tags::FLOAT) => (l.value.i as f64, r.value.f),
            (tags::FLOAT, tags::INTEGER) => (l.value.f, r.value.i as f64),
            (tags::FLOAT, tags::FLOAT) => (l.value.f, r.value.f),
            _ => {
                *o = TValue::new_bool(false);
                return;
            }
        }
    };
    let f = if r == 0.0 || r == -0.0 {
        if l == 0.0 || l == -0.0 {
            -f64::NAN
        } else if l.is_sign_negative() {
            -f64::INFINITY
        } else {
            f64::INFINITY
        }
    } else {
        l / r
    };
    *o = TValue::new_float(f);
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn div(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    let (l, r, o) = bin_math_prefix!(lhs, rhs, out);
    tvalue_div_(l, r, o);
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn floor_div(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    let (l, r, o) = bin_math_prefix!(lhs, rhs, out);
    if r.is_zero() && l.tag == tags::INTEGER {
        *o = TValue::new_bool(false);
        return;
    }
    tvalue_div_(l, r, o);
    let f = match o.tag {
        tags::INTEGER => o.value.i as f64,
        tags::FLOAT => o.value.f,
        _ => return,
    };
    *o = TValue::new_float(f.floor());
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn pow(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    let (l, r, o) = bin_math_prefix!(lhs, rhs, out);
    *o = match (l.tag, r.tag) {
        (tags::INTEGER, tags::INTEGER) => {
            let v = (l.value.i as f64).powf(r.value.i as _);
            TValue::new_float(v)
        }
        (tags::FLOAT, tags::FLOAT) => {
            let v = l.value.f.powf(r.value.f);
            TValue::new_float(v)
        }
        (tags::FLOAT, tags::INTEGER) => {
            let v = l.value.f.powf(r.value.i as f64);
            TValue::new_float(v)
        }
        (tags::INTEGER, tags::FLOAT) => {
            let v = (l.value.i as f64).powf(r.value.f);
            TValue::new_float(v)
        }
        _ => TValue::new_bool(false),
    };
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn neg(lhs: *mut TValue, out: *mut TValue) {
    let (l, o) = (
        crate::get_or_return!(lhs, out),
        crate::get_mut_or_return!(out),
    );
    *o = match l.tag {
        tags::INTEGER => TValue::new_int(l.value.i.wrapping_neg()),
        tags::FLOAT => TValue::new_float(-l.value.f),
        _ => TValue::new_bool(false),
    };
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn bin_and(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    let (l, r, o) = bin_math_prefix!(lhs, rhs, out);
    let i = match (l.tag, r.tag) {
        (tags::INTEGER, tags::INTEGER) => l.value.i & r.value.i,
        _ => {
            *o = TValue::new_bool(false);
            return;
        }
    };
    *o = TValue {
        tag: tags::INTEGER,
        value: TValueInner { i },
    };
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn bin_or(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    let (l, r, o) = bin_math_prefix!(lhs, rhs, out);
    let i = match (l.tag, r.tag) {
        (tags::INTEGER, tags::INTEGER) => l.value.i | r.value.i,
        _ => {
            *o = TValue::new_bool(false);
            return;
        }
    };
    *o = TValue {
        tag: tags::INTEGER,
        value: TValueInner { i },
    };
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn bin_xor(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    let (l, r, o) = bin_math_prefix!(lhs, rhs, out);
    let i = match (l.tag, r.tag) {
        (tags::INTEGER, tags::INTEGER) => l.value.i ^ r.value.i,
        _ => {
            *o = TValue::new_bool(false);
            return;
        }
    };
    *o = TValue {
        tag: tags::INTEGER,
        value: TValueInner { i },
    };
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn bin_rhs(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    let (l, r, o) = bin_math_prefix!(lhs, rhs, out);
    let i = match (l.tag, r.tag) {
        (tags::INTEGER, tags::INTEGER) => l.value.i >> r.value.i,
        _ => {
            *o = TValue::new_bool(false);
            return;
        }
    };
    *o = TValue {
        tag: tags::INTEGER,
        value: TValueInner { i },
    };
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn bin_lhs(lhs: *mut TValue, rhs: *mut TValue, out: *mut TValue) {
    let (l, r, o) = bin_math_prefix!(lhs, rhs, out);
    let i = match (l.tag, r.tag) {
        (tags::INTEGER, tags::INTEGER) => l.value.i << r.value.i,
        _ => {
            *o = TValue::new_bool(false);
            return;
        }
    };
    *o = TValue {
        tag: tags::INTEGER,
        value: TValueInner { i },
    };
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn bin_not(lhs: *mut TValue, out: *mut TValue) {
    let (l, o) = (
        crate::get_or_return!(lhs, out),
        crate::get_mut_or_return!(out),
    );
    let i = match l.tag {
        tags::INTEGER => !l.value.i,
        _ => {
            *o = TValue::new_bool(false);
            return;
        }
    };
    *o = TValue {
        tag: tags::INTEGER,
        value: TValueInner { i },
    };
}

#[runtime_macros::std_tvalue_export(module = "math")]
pub unsafe extern "C" fn to_number(v: *mut TValue) -> f64 {
    let Some(v) = v.as_ref() else {
        return f64::NAN;
    };
    match v.tag {
        tags::BOOLEAN => {
            let v = unsafe { v.value.b };
            if v {
                1.0
            } else {
                0.0
            }
        }
        tags::FLOAT => unsafe { v.value.f },
        tags::INTEGER => {
            let i = { v.value.i };
            i as f64
        }
        tags::STRING_CONST => {
            let s = { v.value.s };
            let slice = unsafe { core::slice::from_raw_parts(s.data, s.len as _) };
            if let Ok(s) = core::str::from_utf8(slice) {
                if let Ok(f) = s.parse::<f64>() {
                    f
                } else {
                    f64::NAN
                }
            } else {
                f64::NAN
            }
        }
        tags::NIL => 0.0,
        _ => unreachable!(),
    }
}

#[cfg(test)]
mod tests {
    use crate::TValue;


    #[test]
    fn int_add() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);

            let expected = lua.load(&format!("{}+{}", l, r)).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::add(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn int_add_edge() {
        let lua = mlua::Lua::new();
        let l = 0;
        let r = -61693221696078929;
        let mut lhs = TValue::new_int(l);
        let mut rhs = TValue::new_int(r);
        let mut result = TValue::new_bool(false);

        let expected = lua.load(&format!("{}+{}", l, r)).eval().map(|v: i64| {
            TValue::new_int(v)
        }).unwrap_or_else(|_| TValue::new_bool(false));

        unsafe {
            crate::math::add(&mut lhs, &mut rhs, &mut result);
        }
        assert_eq!(expected, result);
    }

    #[test]
    fn float_add() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: f64, r: f64)| {
            let mut lhs = TValue::new_float(l);
            let mut rhs = TValue::new_float(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("{}+{}", l, r)).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::add(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn int_neg() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("-({})", l)).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::neg(&mut lhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn float_neg() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: f64)| {
            let mut lhs = TValue::new_float(l);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("-({})", l)).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::neg(&mut lhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn int_sub() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("({})-({})", l, r)).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::sub(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn float_sub() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: f64, r: f64)| {
            let mut lhs = TValue::new_float(l);
            let mut rhs = TValue::new_float(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("({})-({})", l, r)).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::sub(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn int_mul() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("{}*{}", l, r)).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::mul(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn float_mul() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: f64, r: f64)| {
            let mut lhs = TValue::new_float(l);
            let mut rhs = TValue::new_float(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("{}*{}", l, r)).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::mul(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn int_div() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("{}/{}", l, r)).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::div(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn float_div() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: f64, r: f64)| {
            let mut lhs = TValue::new_float(l);
            let mut rhs = TValue::new_float(r);
            let mut result = TValue::new_bool(false);
            unsafe {
                crate::math::div(&mut lhs, &mut rhs, &mut result);
            }
            let expected = lua.load(format!("{}/{}", l, r, )).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));
            if !(expected.is_nan() && result.is_nan()) {
                proptest::prop_assert_eq!(&expected, &result, "{:?} / {:?} != {:?} found {:?}",
                    lhs, rhs, expected, result,
                );
            }
        });
    }

    #[test]
    fn int_floor_div() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("{}//{}", l, r)).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::floor_div(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn float_floor_div() {
        let lua = mlua::Lua::new();

        proptest::proptest!(|(l: f64, r: f64)| {
            let mut lhs = TValue::new_float(l);
            let mut rhs = TValue::new_float(r);
            let mut result = TValue::new_bool(false);

            unsafe {
                crate::math::floor_div(&mut lhs, &mut rhs, &mut result);
            }

            let expected = lua.load(format!("{}//{}", l, r, )).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));
            if !(expected.is_nan() && result.is_nan()) {
                proptest::prop_assert_eq!(&expected, &result, "{:?} // {:?} != {:?} found {:?}",
                    lhs, rhs, expected, result,
                );
            }
        });
    }

    #[test]
    fn int_rem() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("{}%{}", l, r)).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::rem(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(&expected, &result, "{} % {} != {} ({})", lhs, rhs, expected, result);
        });
    }

    #[test]
    fn float_rem() {
        let lua = mlua::Lua::new();

        proptest::proptest!(|(l: f64, r: f64)| {
            let mut lhs = TValue::new_float(l);
            let mut rhs = TValue::new_float(r);
            let mut result = TValue::new_bool(false);

            unsafe {
                crate::math::rem(&mut lhs, &mut rhs, &mut result);
            }

            let expected = lua.load(dbg!(format!("{}%{}", l, r, ))).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));
            if !(expected.is_nan() && result.is_nan()) {
                proptest::prop_assert_eq!(&expected, &result, "{:?} % {:?} != {:?} found {:?}",
                    lhs, rhs, expected, result,
                );
            }
        });
    }

    #[test]
    fn int_exp() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("({})^({})", l, r)).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));

            unsafe {
                crate::math::pow(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(&expected, &result, "{:?} ^ {:?} != {:?} found {:?}", lhs, rhs, expected, result);
        });
    }

    #[test]
    fn float_floor_exp() {
        let lua = mlua::Lua::new();

        proptest::proptest!(|(l: f64, r: f64)| {
            let mut lhs = TValue::new_float(l);
            let mut rhs = TValue::new_float(r);
            let mut result = TValue::new_bool(false);

            unsafe {
                crate::math::pow(&mut lhs, &mut rhs, &mut result);
            }

            let expected = lua.load(format!("({})^({})", l, r, )).eval().map(|v: f64| {
                TValue::new_float(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));
            if !(expected.is_nan() && result.is_nan()) {
                proptest::prop_assert_eq!(&expected, &result, "{:?} ^ {:?} != {:?} found {:?}",
                    lhs, rhs, expected, result,
                );
            }
        });
    }

    #[test]
    fn int_bin_and() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("{l}&{r}")).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));
            unsafe {
                crate::math::bin_and(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(&expected, &result, "{:?} & {:?} != {:?} found {:?}",
                lhs, rhs, expected, result,
            );
        });
    }

    #[test]
    fn int_bin_or() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("{l}|{r}")).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));
            unsafe {
                crate::math::bin_or(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn int_bin_xor() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64, r: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut rhs = TValue::new_int(r);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("{l}~{r}")).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));
            unsafe {
                crate::math::bin_xor(&mut lhs, &mut rhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }

    #[test]
    fn int_bin_not() {
        let lua = mlua::Lua::new();
        proptest::proptest!(|(l: i64)| {
            let mut lhs = TValue::new_int(l);
            let mut result = TValue::new_bool(false);
            let expected = lua.load(&format!("~{l}")).eval().map(|v: i64| {
                TValue::new_int(v)
            }).unwrap_or_else(|_| TValue::new_bool(false));
            unsafe {
                crate::math::bin_not(&mut lhs, &mut result);
            }
            proptest::prop_assert_eq!(expected, result);
        });
    }
}
