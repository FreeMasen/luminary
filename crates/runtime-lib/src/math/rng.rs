//! This module provides PRNG to generate a string of random numbers

use core::sync::atomic::Ordering;
use std::sync::atomic::AtomicI64;

// According to `man srand` if not called the default is 1
static SEED: AtomicI64 = AtomicI64::new(1);
static X: AtomicI64 = AtomicI64::new(0);
static W: AtomicI64 = AtomicI64::new(0);

pub fn seed_random(seed: i64) {
    SEED.store(seed, Ordering::SeqCst);
    X.store(0, Ordering::SeqCst);
    W.store(0, Ordering::SeqCst);
}

pub fn get_random() -> i64 {
    let s = SEED.load(Ordering::Relaxed);
    let old_w = W.load(Ordering::Relaxed);
    let new_w = old_w.wrapping_add(s);
    let old_x = X.load(Ordering::Relaxed);
    let new_x = old_w.wrapping_mul(old_x).wrapping_add(new_w);
    let new_x = new_x >> 32 | new_x << 32;
    W.store(new_w, Ordering::SeqCst);
    X.store(new_x, Ordering::SeqCst);
    new_x
}
