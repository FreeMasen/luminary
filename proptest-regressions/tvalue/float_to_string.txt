# Seeds for failure cases proptest has generated in the past. It is
# automatically read and these particular cases re-run before any
# novel cases are generated.
#
# It is recommended to check this file in to source control so that
# everyone who runs the test benefits from these saved cases.
cc 2999ca4305423c36d735022babf7bca6adae50770867fe8b170a396083c97a84 # shrinks to v = -33439652000.0
cc bf9a2e90ecec098a70e1f5f08d19e985867ab3919d7fec9962e747b4651ed737 # shrinks to v = -31742073000.0
cc 60fc50a0d00a770da84b4ad6534f231c2c2bdf73ff9102787a3f20a2a9e7ce54 # shrinks to v = 31635552000.0
