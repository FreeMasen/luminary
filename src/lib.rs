use core::fmt;
use std::{backtrace::Backtrace, collections::HashMap, path::PathBuf};

use analisar::{
    ast::{BinaryOperator, Expression, Statement},
    Parser,
};
use codegen::CodeGenerator;
use inkwell::{context::Context, module::Module, values::PointerValue};
use tvalue::tvalue_names;
pub mod bytecode;
pub mod codegen;
pub mod symbol_table;
pub mod tvalue;

#[derive(Debug)]
pub enum Error {
    BadFileName(String),
    Other { stack: Backtrace, message: String },
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Error::BadFileName(name) => write!(f, "Bad File Name: {name}"),
            Error::Other { stack, message } => write!(f, "{message}\n{stack}"),
        }
    }
}

impl std::error::Error for Error {}

impl Error {
    pub fn other(msg: impl Into<String>) -> Self {
        Self::Other {
            stack: Backtrace::capture(),
            message: msg.into(),
        }
    }
}

pub fn run_on<'ctx>(context: &'ctx Context, path: PathBuf) -> Module<'ctx> {
    let target_module =
        context.create_module(path.file_stem().expect("file_stem").to_str().unwrap());
    let mut std_builder = tvalue::TValueModuleBuilder::new(&context);
    let tvalue_module = std_builder.gen_lib();
    target_module
        .link_in_module(tvalue_module)
        .expect("linking should work");
    let generator = codegen::CodeGenerator::new(target_module);
    let lua = std::fs::read_to_string(&path)
        .unwrap_or_else(|e| panic!("Error reading {}: {e}", path.display()));
    emit_code(&generator, lua);
    generator.into_module()
}

pub fn emit_code<'ctx>(gen: &CodeGenerator<'ctx>, lua: String) {
    let mut variables = HashMap::new();
    let mut p = Parser::new(lua.as_bytes());
    gen.emit_main_and_move_to_entry();
    while let Some(stmt) = p.next() {
        let stmt = stmt.unwrap();
        match stmt {
            Statement::Assignment {
                local: _,
                targets,
                values,
            } => {
                for (idx, target) in targets.iter().enumerate() {
                    let value = values.get(idx);
                    emit_assignment(&gen, target, value, &mut variables);
                }
            }
            Statement::Expression(Expression::FuncCall(call)) => {
                let name = expression_to_name(&call.prefix);
                if name != "print" {
                    panic!("Invalid function name, only `print` is supported");
                }
                match call.args {
                    analisar::ast::Args::ExpList(exprs) => {
                        if exprs.is_empty() {
                            let ptr = expression_to_ptr(
                                &gen,
                                &Expression::Nil,
                                "_",
                                &mut variables,
                            );
                            gen.perform_print(std::iter::once(ptr));
                            continue;
                        }
                        gen.perform_print(exprs.iter().map(|e| expression_to_ptr(&gen, e, "_", &mut variables)));
                    }
                    analisar::ast::Args::Table(_) => panic!("tables unsupported"),
                    analisar::ast::Args::String(lit) => {
                        let ptr = gen.init_tvalue_string(&lit.0, "_");
                        gen.perform_print(std::iter::once(ptr));
                    }
                };
            }
            Statement::Expression(Expression::BinOp { left, op, right }) => {
                use tvalue_names::math::*;
                let op_name = match op {
                    BinaryOperator::Add => ADD,
                    BinaryOperator::Subtract => SUB,
                    BinaryOperator::Multiply => MUL,
                    BinaryOperator::Divide => DIV,
                    BinaryOperator::FloorDivide => FLOOR_DIV,
                    BinaryOperator::Power => POW,
                    BinaryOperator::Modulo => MOD,
                    BinaryOperator::BitwiseAnd => AND,
                    BinaryOperator::BitwiseOr => OR,
                    BinaryOperator::RightShift => RSH,
                    BinaryOperator::LeftShift => LSH,
                    _ => unimplemented!("{op:?}"),
                };
                let lhs = expression_to_ptr(&gen, &*left, "lhs", &mut variables);
                let rhs = expression_to_ptr(&gen, &*right, "rhs", &mut variables);
                let dest = gen.alloca_tvalue("result");
                let _success = gen.perform_binary_op("success", op_name, lhs, rhs, dest);
            }
            _ => unimplemented!("{stmt:?}"),
        }
    }
    gen.emit_main_return();
}

fn emit_assignment<'ctx>(
    gen: &CodeGenerator<'ctx>,
    target: &Expression,
    init: Option<&Expression>,
    vars: &mut HashMap<String, PointerValue<'ctx>>,
) -> PointerValue<'ctx> {
    let name = expression_to_name(target);
    let Some(init) = init else {
        let ptr = gen.alloca_tvalue(&name);
        vars.insert(name, ptr);
        return ptr;
    };
    let ptr = expression_to_ptr(gen, init, &name, vars);
    vars.insert(name, ptr);
    ptr
}

pub fn expression_to_name(expr: &Expression) -> String {
    match expr {
        Expression::Name(n) => n.name.to_string(),
        _ => unimplemented!("expr_to_name: {expr:?}"),
    }
}

fn expression_to_ptr<'ctx>(
    gen: &CodeGenerator<'ctx>,
    expr: &Expression,
    name_if_const: &str,
    vars: &mut HashMap<String, PointerValue<'ctx>>,
) -> PointerValue<'ctx> {
    match expr {
        Expression::Nil => gen.alloca_tvalue(&name_if_const),
        Expression::False => gen.init_tvalue_bool(false, &name_if_const),
        Expression::True => gen.init_tvalue_bool(true, &name_if_const),
        Expression::Numeral(n) => {
            let float: f32 = n.0.parse().expect("float value");
            gen.init_tvalue_num(float, &name_if_const)
        }
        Expression::LiteralString(s) => gen.init_tvalue_string(&s.0, &name_if_const),
        Expression::Name(n) => {
            let Some(ptr) = vars.get(&n.name[..]) else {
                panic!("unknown variable {n:?}");
            };
            *ptr
        }
        Expression::BinOp { left, op, right } => emit_bin_op(gen, left, right, *op, vars),
        _ => unimplemented!("expression_to_ptr: {expr:?}"),
    }
}

fn emit_bin_op<'ctx>(
    gen: &CodeGenerator<'ctx>,
    left: &Expression,
    right: &Expression,
    op: BinaryOperator,
    vars: &mut HashMap<String, PointerValue<'ctx>>,
) -> PointerValue<'ctx> {
    use tvalue_names::math::*;
    let op_name = match op {
        BinaryOperator::Add => ADD,
        BinaryOperator::Subtract => SUB,
        BinaryOperator::Multiply => MUL,
        BinaryOperator::Divide => DIV,
        BinaryOperator::FloorDivide => FLOOR_DIV,
        BinaryOperator::Power => POW,
        BinaryOperator::Modulo => MOD,
        BinaryOperator::BitwiseAnd => AND,
        BinaryOperator::BitwiseOr => OR,
        BinaryOperator::RightShift => RSH,
        BinaryOperator::LeftShift => LSH,
        _ => unimplemented!("{op:?}"),
    };
    let lhs = expression_to_ptr(&gen, &*left, "lhs", vars);
    let rhs = expression_to_ptr(&gen, &*right, "rhs", vars);
    let dest = gen.alloca_tvalue("result");
    let _success = gen.perform_binary_op("success", op_name, lhs, rhs, dest);
    dest
}
