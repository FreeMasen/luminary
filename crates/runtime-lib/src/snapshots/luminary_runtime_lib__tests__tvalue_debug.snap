---
source: src/lib.rs
expression: "&[TValue::new_nil(), TValue::new_bool(false), TValue::new_bool(true),\n            TValue::new_int(42), TValue::new_float(42.1),\n            TValue::new_str(\"hello world!\")]"
---
[
    TValue::::Nil,
    TValue::Bool(
        false,
    ),
    TValue::Bool(
        true,
    ),
    TValue::Int(
        42,
    ),
    TValue::Float(
        42.1,
    ),
    TValue::StringConst(
        "hello world!",
    ),
]
